Add SEMS Portal integration

andrewgierens opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This PR is to add a new integration for the SEMS portal and the devices registered. The data from this portal is energy related data for the GoodWe inverters.

_Why this and not the Goodwe local integration?_ 
I'd rather use the Goodwe integration, but currently it does not support Homekit sensors (used to measure grid, home and pv load)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: NA
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29718

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

andrewgierens committed:
Init commit with live data

zweckj commented on homeassistant/components/sems_portal/sems_login.py:
```diff
@@ -0,0 +1,20 @@
+from aiohttp import ClientSession
+
+
+async def login_to_sems(session: ClientSession, account: str, pwd: str) -> Any:
```
needs to move to external PyPi package (https://developers.home-assistant.io/docs/development_checklist/)

zweckj commented on homeassistant/components/sems_portal/__init__.py:
```diff
@@ -0,0 +1,165 @@
+        ) as error:
+            raise UpdateFailed(error) from error
+
+        data = {
```
should probably also move to pypi package

andrewgierens committed:
Removed unused properties
Removed comments for mapping
Moved sems portal logic to pypi package
Added pypi package to requirements

andrewgierens committed:
Merge branch 'dev' into sems_portal

andrewgierens committed:
Removed mapping and replaced with collated func

andrewgierens committed:
Removed 'bettery' word from spelling ignore list

andrewgierens committed:
Fixed auto updating
Added override for strict types
Added override for coverage

andrewgierens committed:
Merge branch 'dev' into sems_portal

zweckj commented on homeassistant/components/sems_portal/__init__.py:
```diff
@@ -0,0 +1,107 @@
+    return True
+
+
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
What do you use that update_listener for? You don't seem to have an options flow

andrewgierens replied:
Thanks for that, that was some left over from a copy paste

zweckj commented on homeassistant/components/sems_portal/sensor.py:
```diff
@@ -0,0 +1,278 @@
+from . import DOMAIN, SemsDataUpdateCoordinator
+
+
+class SemsInformationSensor(SensorEntity):
```
Will you add more Platforms later? Then I'd definitely add a common base class, if not I'd still think about it to reduce some of the duplicate code in the inits.

andrewgierens replied:
Have added a common base class in entity.py (Saw other integrations did this, lmk if this isn't correct practice)

zweckj commented on homeassistant/components/sems_portal/sensor.py:
```diff
@@ -0,0 +1,278 @@
+        """Return the state of the sensor."""
+        return self.coordinator.data["powerPlant"]["info"][self.entity_description.key]
+
+    async def async_added_to_hass(self) -> None:
+        """Connect to dispatcher listening for entity data notifications."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_update(self) -> None:
+        """Get the latest data from OWM and updates the states."""
+        await self.coordinator.async_request_refresh()
+
```
I believe you don't need those if you also inherit from `CoordinatorEntity`

zweckj commented on homeassistant/components/sems_portal/sensor.py:
```diff
@@ -0,0 +1,278 @@
+            self.entity_description.key
+        ]
+
+    async def async_added_to_hass(self) -> None:
+        """Connect to dispatcher listening for entity data notifications."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_update(self) -> None:
+        """Get the latest data from OWM and updates the states."""
+        await self.coordinator.async_request_refresh()
```
same here

zweckj commented on homeassistant/components/sems_portal/sensor.py:
```diff
@@ -0,0 +1,278 @@
+
+SENSOR_TYPES_POWERSTATION: tuple[SensorEntityDescription, ...] = (
+    SensorEntityDescription(
+        name="Powerstation Id",
```
is it possible to use `translation_key` here and move the nae to `strings.json`? Would allow for better localization support (applies to all sensors)

andrewgierens committed:
Moved to sems instead of sems_portal

andrewgierens committed:
Removed unused config sync
Updated sensor to use a base class
Updated base class to use CoordinatorEntity
Updated sensors to use translated text

andrewgierens committed:
Split out base class
Added SensorEntity to base class (preventing updating)

andrewgierens committed:
Merge branch 'dev' into sems_portal

andrewgierens committed:
Fixed up doco path

zweckj commented on homeassistant/components/sems/manifest.json:
```diff
@@ -0,0 +1,11 @@
+  "codeowners": ["@andrewgierens"],
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/sems",
+  "homekit": {},
```
sorry missed that one: please remove that empty key

andrewgierens replied:
No worries - Done :)

andrewgierens committed:
Removed empty key

andrewgierens committed:
Merge branch 'dev' into sems_portal

zweckj reviewed (approved):
Looks good to me - but I don't have permissions to approve 😅

andrewgierens committed:
Merge branch 'dev' into sems_portal

andrewgierens committed:
Merge branch 'dev' into sems_portal

andrewgierens committed:
Merge branch 'dev' into sems_portal

andrewgierens committed:
Merge branch 'dev' into sems_portal
